#Exercise 1: Merge ATO dataset with Postcode data
In this exercise, we will learn how to merge the ATO dataset with the postcode dataset.

The Australian Taxation Office (ATO) and postcode datasets can be found in the Packt GitHub repository here: 
- https://github.com/PacktWorkshops/The-Data-Science-Workshop/blob/master/Chapter12/Dataset/taxstats2015.csv
- https://github.com/PacktWorkshops/The-Data-Science-Workshop/blob/master/Chapter12/Dataset/taxstats2016individual06taxablestatusstateterritorypostcodetaxableincome%20(2).xlsx


The source of the dataset can be found here: 
- https://data.gov.au/dataset/ds-dga-5c99cfed-254d-40a6-af1c-47412b7de6fe/details.
- https://data.gov.au/data/dataset/5fa69f19-ec44-4c46-88eb-0f6fd5c2f43b/resource/31456a8b-7cd9-4f55-bdcb-2acbbb209bd0/download/taxstats2016individual06taxablestatusstateterritorypostcodetaxableincome.xlsx



1. Open on a new Colab notebook and import the pandas package

In [0]:
import pandas as pd

2. Assign the link to the ATO dataset to a variable called 'file_url'




In [0]:
file_url = 'https://raw.githubusercontent.com/PacktWorkshops/The-Data-Science-Workshop/master/Chapter12/Dataset/taxstats2015.csv'

3. Using the read_csv .method() from the package pandas, load the dataset into a new DataFrame called 'df'

In [0]:
df = pd.read_csv(file_url)

4. Display the dimensions of this DataFrame using the .shape attribute

In [4]:
df.shape

(2473, 28)

The ATO dataset contains 2471 rows and 28 columns.

5. Display the first 5 rows of the ATO DataFrame using the .head() method

In [5]:
df.head()

Postcode  ...  Median super total accounts balance
0      2000  ...                                 1110
1      2006  ...                                15544
2      2007  ...                                 5225
3      2008  ...                                10373
4      2009  ...                                20494

[5 rows x 28 columns]

The first column called 'Postcode' is the one we will use to merge with the second dataset. Postcode are the name in Australia for zip code. It is an identifier for postal areas. We are interested to learn more about each of this postcode. Let's make sure they are all unique in this dataset.

6. Display the number of unique values for the variable 'Postcode' using the .nunique() method

In [6]:
df['Postcode'].nunique()

2473

There are 2473 unique values in this column and the DataFrame has 2473 rows so we are sure the variable 'Postcode' contains only unique values.

7. Assign the link to the Postcode dataset to a variable called 'postcode_df'




In [0]:
postcode_url = 'https://github.com/PacktWorkshops/The-Data-Science-Workshop/blob/master/Chapter12/Dataset/taxstats2016individual06taxablestatusstateterritorypostcodetaxableincome%20(2).xlsx?raw=true'

8. Load the Postcode dataset into a new DataFrame called postcode_df using the .read_excel() method. We will only load the sheet 'Individuals Table 6B' so we ned to provide this name to the parameter 'sheet_name'. Also the header column in this spreadsheet is located at the third row so we need to specify it to the parameter header (don't forget index starts with 0 in python).

In [0]:
postcode_df = pd.read_excel(postcode_url, sheet_name='Individuals Table 6B', header=2)

9. Print the dimensions of postcode_df using the .shape attribute

In [9]:
postcode_df.shape

(2567, 150)

This DataFrame contains 2567 rows for 150 columns. By merging it with the ATO dataset, we will get additional information for each postcode.

10. Print the first 5 rows of postcode_df using the .head() method

In [10]:
postcode_df.head()

State/ Territory1  ... Excess private health reduction or refund (rebate reduced) \n$
0               ACT  ...                                             854227            
1               ACT  ...                                             173475            
2               ACT  ...                                            1381661            
3               ACT  ...                                            1004593            
4               ACT  ...                                             754097            

[5 rows x 150 columns]

We can see the second column contains the postcode value and this is the one we will use to merge on with the ATO dataset. Let's check they are unique.

11. Print the number of unique values in this column using the .nunique() method

In [11]:
postcode_df['Postcode'].nunique()

2567

There are 2567 unique values and this corresponds exactly to the number of rows of this DataFrame so we absolutely sure this column contains unique values. This also means after merging the 2 tables, there will be only one to one match. We won't have the case where we will get multiple rows from one of the dataset matching with only one rows of the other one.

12. Perform a left join on the 2 DataFrames using the .merge() method and save the results into a new DataFrame called merged_df. Specify the parameter how='left' and on='Postcode'.

In [0]:
merged_df = pd.merge(df, postcode_df, how='left', on='Postcode')

13. Print the dimensions of the new merged DataFrame using the .shape attribute

In [13]:
merged_df.shape

(2473, 177)

We got exactly 2473 rows after merging which is what we expect as we used a left join and there were 1-to-1 match on the 'Postcode' column from both orginal DataFrame. Also we have now 177 columns which is the objective of this exercise. But before concluding it, we want to see there are any postcode that didn't match between the 2 datasets. To do so, we will be looking at one column from the right hand side DataFrame (the Postcode dataset) and see if there is any missing value.

14. Print the total number of missing values from the column 'State/Territory1' by combining the .isna() and .sum() methods

In [14]:
merged_df['State/ Territory1'].isna().sum()

4

There are 4 postcode from the ATO dataset that didn't match with the Postcode code. Let's see which ones they are.

15. Print the missing postcodes using the .iloc() method

In [15]:
merged_df.loc[merged_df['State/ Territory1'].isna(), 'Postcode']

631     3010
1494    4462
2072    6068
2332    6758
Name: Postcode, dtype: object

The missing postcodes from the Postcode dataset are 3010, 4462, 6068 and 6758. In a real project, you will have to get in touch with your stakeholders or the data team to see if you are able to get these data.

Excellent! We have successfully merged the 2 datasets of interest and have expanded the number of features. We now have a much richer dataset and will be able to perform a more detailed analysis on it.